lib/types-simple.nix

```diff
@@ -0,0 +1,457 @@
+      then reqfs == vfs
+        # all fields have to exist in the value
+        # reqfs - vfs
+      else lib.subtractLists vfs reqfs == []
+        # whithout req, only opt fields must be in the value
+        # (vfs - reqfs) - otfs
+        && lib.subtractLists optfs (lib.subtractLists reqfs vfs) == []);
+    variant = variants.product;
+    extraFields = {
+      inherit opt req;
+    };
+  };
+
+  # sum type with alternatives of the specified types
+  # sum { left = string; right = bool; }:
+  #   { left = "work it"; }
+  #   { right = false; }
+  # sum { true = unit; false = unit; } <- that’s isomorphic to bool
+  #   { true = {}; }
+  #   { false = {}; }
+  # sum { X = product { name = string; age = int; }; Y = list unit; }
+  #   { X = { name = "peter shaw"; age = 22; }; }
+  #   { Y = [ {} {} {} {} {} {} {} {} ]; }
+  sum = alts: assert alts != {}; mkBaseType {
+    name = "sum";
+    description = "< " +
+      lib.concatStringsSep " | "
+        (lib.mapAttrsToList (n: t: "${n}: ${describe t}") alts)
+      + " >";
+    check = v:
+      let alt = builtins.attrNames v;
+      in builtins.isAttrs v
+      # exactly one of the alts has to be used by the value
+      && builtins.length alt == 1
+      # the alt tag of the value should of course be a possibility
+      && alts ? ${lib.head alt};
+    variant = variants.sum;
+    extraFields = {
+      inherit alts;
+    };
+  };
+
+  # untagged union type
+  # ATTENTION: this leads to *bad* type checker errors in practice,
+  # you also can’t do pattern matching; use sum if possible.
+  # union [ bool int ]
+  #   3
+  #   true
+  # list (union [ int string ])
+  #   [ "foo" 34 "bar" ]
+  # please don’t use this.
+  union = altList: assert altList != []; mkBaseType {
+    name = "union";
+    description = "one of [ "
+      + lib.concatMapStringsSep ", " describe altList
+      + " ]";
+    # any type that checks out is fine
+    check = v: lib.any (t: t.check v) altList;
+    variant = variants.union;
+    extraFields = {
+      inherit altList;
+    };
+  };
+
+  # TODO: should scalars be allowed as nest types?
+  # TODO: how to implement?
+  # nested = nest: t: mkBaseType {
+  #   description = "nested ${describe nest} of ${describe t}";
+  #   check = nest.check ;
+  #   variant = nest.variant;
+  #   extraFields = {
+
+
+  ## -- FUNCTIONS --
+
+  # TODO: pattern match function
+  # match =
+
+  # Default values for some types, chosen pretty arbitrarily.
+  # Can be used to populate products which might have less fields
+  # than originally specified:
+  # let t = product { foo : string, bar = product { baz = int; }; }
+  #     def = defaults t; # { foo = ""; bar.baz = 0; };
+  #     val = { foo = "hello"; };
+  # in checkType t (defaults t // val)
+  # or even recursiveUpdate if you want to be naughty.
+  # Hint: It’s probably better to use `productOpt` instead,
+  #       but more power to the people.
+  defaults = t:
```

Comment:
Hm, I don't think this has any more benefit to having to specify the defaults directly, especially if it comes to code readability. You need to look up the default here instead of just following the code you're already looking at.

Reply:
I implemented that as a crutch for optional fields originally, but `productOpt` kind of replaced it pretty quickly. It might make sense to remove it again, yes.